sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java

```diff
@@ -52,280 +59,491 @@
 
-    final PCollection<String> pCollection = pCollection();
-    PAssert.that(pCollection).containsInAnyOrder(DUMMY);
-    pipeline.run().waitUntilFinish();
+      @Category(NeedsRunner.class)
+      @Test
+      public void testConflict_needsRunnerAnnotation_expectIllegalStateException()
+          throws Exception {
 
-    // dangling PAssert
-    PAssert.that(pCollection).containsInAnyOrder(DUMMY);
-  }
+        addTransform(pCollection(pipeline));
 
-  @Test
-  public void testPipelineHasAbandonedPTransformNode() throws Throwable {
-    exception.expect(TestPipeline.AbandonedNodeException.class);
-    exception.expectMessage("PTransform");
+        // @NeedsRunner annotation with a CrashingRunner indicates a configuration issue
+      }
 
-    final PCollection<String> pCollection = pCollection();
-    PAssert.that(pCollection).containsInAnyOrder(DUMMY);
-    pipeline.run().waitUntilFinish();
+      @Category(RunnableOnService.class)
+      @Test
+      public void testConflict_runnableOnServiceAnnotation_expectIllegalStateException()
+          throws Exception {
 
-    // dangling PTransform
-    addTransform(pCollection);
-  }
+        addTransform(pCollection(pipeline));
 
-  @Test
-  public void testNormalFlowWithPAssert() throws Throwable {
-    PAssert.that(pCollection()).containsInAnyOrder(DUMMY);
-    pipeline.run().waitUntilFinish();
-  }
+        // @RunnableOnService annotation with a CrashingRunner indicates a configuration issue
+      }
 
-  @Test
-  public void testAutoAddMissingRunFlow() throws Throwable {
-    PAssert.that(pCollection()).containsInAnyOrder(DUMMY);
-    // missing pipeline#run, but have it auto-added.
-    pipeline.enableAutoRunIfMissing(true);
-  }
+    }
+
+    private static List<Object[]> extractTests(final Class<?> testClass) {
+      final ArrayList<org.junit.runner.Description> testDescriptions = Request.aClass(testClass)
+                                                                              .getRunner()
+                                                                              .getDescription()
+                                                                              .getChildren();
+
+      return FluentIterable.from(testDescriptions)
+                           .transform(new Function<org.junit.runner.Description, Object[]>() {
+
+                             @Override
+                             public Object[] apply(final org.junit.runner.Description description) {
+                               return new Object[] {
+                                   Request.method(testClass, description.getMethodName()),
+                                   description.getTestClass().getSimpleName() + "#"
+                                       + description.getMethodName()
+                               };
+                             }
+                           })
+                           .toList();
+    }
+
+    private TestRule withPossibleExceptions(final String testName) {
```

Comment:
This worries me.

I'm ok with having the tests run with `Parameterized`, but if we do so we should still be able to determine which case we're in and set the expected exception per-test within the test based on the value of the parameters, rather than inspecting the name of the test. For example, `testConflict_runnableOnServiceAnnotation_expectIllegalStateException` doesn't demonstrate by reading the test code directly that it expects an exception.